packages/jsts/src/analysis/projectAnalysis/tsconfigCache.ts

```diff
@@ -0,0 +1,119 @@
+ * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the Sonar Source-Available License for more details.
+ *
+ * You should have received a copy of the Sonar Source-Available License
+ * along with this program; if not, see https://sonarsource.com/license/ssal/
+ */
+
+import { dirname } from 'path/posix';
+import { debug, info } from '../../../../shared/src/helpers/logging.js';
+import { createProgramOptions } from '../../program/program.js';
+import { UNINITIALIZED_ERROR } from './tsconfigs.js';
+
+export class Cache {
+  private readonly inputFileToTsConfigFilesMap = new Map<string, string | null>();
+  public discoveredTsConfigFiles = new Set<string>();
+  public key: string | undefined;
+  public originalTsConfigFiles: string[] = [];
+  private pendingTsConfigFiles: string[] = [];
+  public initialized = false;
+
+  getTsConfigForInputFile(inputFile: string) {
+    if (!this.initialized) {
+      throw new Error(UNINITIALIZED_ERROR);
+    }
+    if (!this.inputFileToTsConfigFilesMap.has(inputFile)) {
+      this.findTsConfigForInputFile(inputFile);
+    }
+
+    return this.inputFileToTsConfigFilesMap.get(inputFile) ?? null;
+  }
+
+  findTsConfigForInputFile(inputFile: string) {
+    this.pendingTsConfigFiles = this.improvedPendingTsConfigOrder(inputFile);
+
+    debug(`Continuing BFS for file: ${inputFile}, pending order: ${this.pendingTsConfigFiles}`);
+    while (this.pendingTsConfigFiles.length) {
+      const tsConfigPath = this.pendingTsConfigFiles.pop()!;
+      debug(`Computing tsconfig ${tsConfigPath} from bridge`);
+      const tsConfigFile = createProgramOptions(tsConfigPath);
+      tsConfigFile.rootNames.forEach(file => {
+        if (!this.inputFileToTsConfigFilesMap.has(file)) {
+          this.inputFileToTsConfigFilesMap.set(file, tsConfigPath);
+        }
+      });
+      if (tsConfigFile.projectReferences?.length) {
+        info(
+          `Adding referenced project's tsconfigs ${tsConfigFile.projectReferences.map(ref => ref.path)}`,
+        );
+        tsConfigFile.projectReferences
+          .filter(refPath => !this.discoveredTsConfigFiles.has(refPath.path))
+          .forEach(refPath => {
+            this.discoveredTsConfigFiles.add(refPath.path);
+            this.pendingTsConfigFiles.unshift(refPath.path);
+          });
+      }
+      if (this.inputFileToTsConfigFilesMap.has(inputFile)) {
+        const foundTsConfigFile = this.inputFileToTsConfigFilesMap.get(inputFile);
+        info(
+          `Using tsConfig ${foundTsConfigFile} for file source file ${inputFile} (${this.pendingTsConfigFiles.length}/${this.discoveredTsConfigFiles.size} tsconfigs not yet checked)`,
+        );
+        return this.inputFileToTsConfigFilesMap.get(inputFile);
```

Comment:
no need to get it again, no?
```suggestion
        return foundTsConfigFile;
```